lib/executor/variable_arrival_rate.go

```diff
@@ -271,59 +170,74 @@ func (varc VariableArrivalRateConfig) HasWork(et *lib.ExecutionTuple) bool {
 //TODO: combine with the ConstantArrivalRate?
 type VariableArrivalRate struct {
 	*BaseExecutor
-	config             VariableArrivalRateConfig
-	plannedRateChanges []rateChange
+	config VariableArrivalRateConfig
 }
 
 // Make sure we implement the lib.Executor interface.
 var _ lib.Executor = &VariableArrivalRate{}
 
-// streamRateChanges is a helper method that emits rate change events at their
-// proper time.
-func (varr VariableArrivalRate) streamRateChanges(ctx context.Context, startTime time.Time) <-chan rateChange {
-	ch := make(chan rateChange)
-	go func() {
-		for _, step := range varr.plannedRateChanges {
-			offsetDiff := step.timeOffset - time.Since(startTime)
-			if offsetDiff > 0 { // wait until time of event arrives
-				select {
-				case <-ctx.Done():
-					return // exit if context is cancelled
-				case <-time.After(offsetDiff): //TODO: reuse a timer?
-					// do nothing
-				}
+func (varc VariableArrivalRateConfig) cal(et *lib.ExecutionTuple, ch chan<- time.Duration) {
+	// TODO: add inline comments with explanation of what is happening
+	// for now just link to https://github.com/loadimpact/k6/issues/1299#issuecomment-575661084
+	// TODO: rewrite this as state based functions/lambdas or something, drop the channel and let the user call
+	// it and precalculate the values, something like the `next` below
+	start, offsets, _ := et.GetStripedOffsets(et.ES)
+	li := -1
+	// TODO: move this to a utility function, or directly what GetStripedOffsets uses once we see everywhere we will use it
+	next := func() int64 {
+		li++
+		return offsets[li%len(offsets)]
+	}
+	defer close(ch) // TODO: maybe this is not a good design - closing a channel we get
+	var (
+		stageStart                   time.Duration
+		timeUnit                     = time.Duration(varc.TimeUnit.Duration).Nanoseconds() // TODO: test
+		doneSoFar, endCount, to, dur float64
+		from                         = float64(varc.StartRate.ValueOrZero()) / float64(timeUnit)
+		i                            = float64(start + 1)
```

Comment:
This calculates full iterations by calculating the area "under the graph". But the striping algorithm starts from 0 so the iteration with index 0 will not be calculated correctly, or at least it will be calculated as from time 0 . Then if you have a constant stage from 2 to 2 over 5 seconds instead of 10 iterations we will start 11 (as 1 will be started at 0 and 1 and the 5second mark).

Reply:
resolved in d68ebe3